File: backend-cs/Services/EmailNotificationService.cs
using System.Text.Json;
using MailKit.Net.Smtp;
using MailKit.Security;
using MimeKit;
using DriveChill.Models;

namespace DriveChill.Services;

/// <summary>
/// Sends alert notification emails via SMTP using MailKit.
/// Supports STARTTLS (port 587) via use_tls and implicit TLS (port 465) via use_ssl.
/// Settings (host, port, credentials, recipients) are read from the DB at send time,
/// so runtime changes take effect on the next alert without a restart.
/// </summary>
public class EmailNotificationService
{
    private readonly DbService _db;
    private readonly ILogger<EmailNotificationService> _log;

    // ── Integration health tracking ──────────────────────────────────────
    private long _successCount;
    private long _failureCount;
    private DateTimeOffset? _lastSentAt;
    private string? _lastError;

    public long SuccessCount => Interlocked.Read(ref _successCount);
    public long FailureCount => Interlocked.Read(ref _failureCount);
    public DateTimeOffset? LastSentAt => _lastSentAt;
    public string? LastError => _lastError;

    public EmailNotificationService(DbService db, ILogger<EmailNotificationService> log)
    {
        _db  = db;
        _log = log;
    }

    /// <summary>
    /// Send an alert notification email. Silently no-ops when email is disabled or unconfigured.
    /// </summary>
    public async Task SendAlertAsync(AlertEvent evt, CancellationToken ct = default)
    {
        var s = await _db.GetEmailSettingsAsync(ct);
        if (!s.Enabled || string.IsNullOrEmpty(s.SmtpHost))
            return;

        var recipients = JsonSerializer.Deserialize<string[]>(s.RecipientList)
                         ?? Array.Empty<string>();
        if (recipients.Length == 0)
            return;

        var password = await _db.GetSmtpPasswordAsync(ct);

        try
        {
            var subject = $"DriveChill Alert: {evt.SensorName} {evt.Condition} {evt.Threshold}";
            var body    = FormatBody(evt);

            var message = BuildMessage(s.SenderAddress, recipients, subject, body, isHtml: false);
            await SendViaMailKitAsync(s, password, message, ct);
            _lastSentAt = DateTimeOffset.UtcNow;
            Interlocked.Increment(ref _successCount);
            _lastError = null;
            _log.LogInformation("Alert email sent for rule {RuleId}", evt.RuleId);
        }
        catch (Exception ex)
        {
            Interlocked.Increment(ref _failureCount);
            _lastError = ex.Message;
            _log.LogWarning(ex, "Email delivery failed: {Service} rule={RuleId} smtp_host={SmtpHost}",
                "email", evt.RuleId, s.SmtpHost);
        }
    }

    /// <summary>
    /// Send a test email to verify SMTP configuration. Returns null on success, error message on failure.
    /// </summary>
    public async Task<string?> SendTestAsync(CancellationToken ct = default)
    {
        var s = await _db.GetEmailSettingsAsync(ct);
        if (!s.Enabled || string.IsNullOrEmpty(s.SmtpHost))
            return "Email notifications are not configured.";

        var recipients = JsonSerializer.Deserialize<string[]>(s.RecipientList)
                         ?? Array.Empty<string>();
        if (recipients.Length == 0)
            return "No recipients configured.";

        var password = await _db.GetSmtpPasswordAsync(ct);

        try
        {
            var message = BuildMessage(s.SenderAddress, recipients,
                "DriveChill Test Notification",
                "This is a test email from DriveChill to confirm that email notifications are working.",
                isHtml: false);
            await SendViaMailKitAsync(s, password, message, ct);
            _lastSentAt = DateTimeOffset.UtcNow;
            Interlocked.Increment(ref _successCount);
            _lastError = null;
            return null;   // success
        }
        catch (Exception ex)
        {
            Interlocked.Increment(ref _failureCount);
            _lastError = ex.Message;
            return ex.Message;
        }
    }

    /// <summary>
    /// Send an HTML analytics report email. Returns true on successful delivery.
    /// </summary>
    public virtual async Task<bool> SendHtmlReportAsync(string subject, string htmlBody, CancellationToken ct = default)
    {
        var s = await _db.GetEmailSettingsAsync(ct);
        if (!s.Enabled || string.IsNullOrEmpty(s.SmtpHost))
            return false;

        var recipients = JsonSerializer.Deserialize<string[]>(s.RecipientList)
                         ?? Array.Empty<string>();
        if (recipients.Length == 0)
            return false;

        var password = await _db.GetSmtpPasswordAsync(ct);

        try
        {
            var message = BuildMessage(s.SenderAddress, recipients, subject, htmlBody, isHtml: true);
            await SendViaMailKitAsync(s, password, message, ct);
            _lastSentAt = DateTimeOffset.UtcNow;
            Interlocked.Increment(ref _successCount);
            _lastError = null;
            return true;
        }
        catch (Exception ex)
        {
            Interlocked.Increment(ref _failureCount);
            _lastError = ex.Message;
            _log.LogWarning(ex, "Email delivery failed: {Service} subject={Subject} smtp_host={SmtpHost}",
                "email", subject, s.SmtpHost);
            return false;
        }
    }

    // -----------------------------------------------------------------------

    private static MimeMessage BuildMessage(string sender, string[] recipients, string subject, string body, bool isHtml)
    {
        var message = new MimeMessage();
        message.From.Add(MailboxAddress.Parse(sender));
        foreach (var r in recipients)
            message.To.Add(MailboxAddress.Parse(r));
        message.Subject = subject;

        message.Body = isHtml
            ? new TextPart("html")  { Text = body }
            : new TextPart("plain") { Text = body };

        return message;
    }

    /// <summary>
    /// Connects to the SMTP server using MailKit with the correct security option:
    ///   use_ssl  = true  → SecureSocketOptions.SslOnConnect  (implicit TLS, typically port 465)
    ///   use_tls  = true  → SecureSocketOptions.StartTls      (STARTTLS, typically port 587)
    ///   neither          → SecureSocketOptions.Auto           (let MailKit negotiate)
    /// </summary>
    private static async Task SendViaMailKitAsync(
        EmailNotificationSettingsRecord s, string password, MimeMessage message, CancellationToken ct)
    {
        SecureSocketOptions secureOption;
        if (s.UseSsl)
            secureOption = SecureSocketOptions.SslOnConnect;
        else if (s.UseTls)
            secureOption = SecureSocketOptions.StartTls;
        else
            secureOption = SecureSocketOptions.Auto;

        using var client = new SmtpClient();
        client.Timeout = 15_000;
        await client.ConnectAsync(s.SmtpHost, s.SmtpPort, secureOption, ct);

        if (!string.IsNullOrEmpty(s.SmtpUsername))
            await client.AuthenticateAsync(s.SmtpUsername, password, ct);

        await client.SendAsync(message, ct);
        await client.DisconnectAsync(quit: true, ct);
    }

    private static string FormatBody(AlertEvent evt)
    {
        var direction = evt.Condition == "above" ? "exceeded" : "fell below";
        return $"""
            DriveChill alert triggered.

            Sensor  : {evt.SensorName}
            Value   : {evt.ActualValue:F1}
            Threshold: {direction} {evt.Threshold:F1}
            Fired at: {evt.FiredAt:u}

            {(string.IsNullOrWhiteSpace(evt.Message) ? "" : $"Note: {evt.Message}")}
            """;
    }
}

File: backend/app/services/email_notification_service.py
"""Email notification service for DriveChill alert delivery."""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from email.message import EmailMessage

try:
    import aiosmtplib
except ImportError:
    aiosmtplib = None  # type: ignore

from app.db.repositories.email_notification_repo import EmailNotificationRepo

logger = logging.getLogger(__name__)


class EmailNotificationService:
    def __init__(self, repo: EmailNotificationRepo) -> None:
        self._repo = repo

    async def send_alert(
        self,
        sensor_name: str,
        value: float,
        threshold: float,
        unit: str = "\u00b0C",
    ) -> int:
        """Send an alert email to all configured recipients.

        Returns the number of messages successfully sent (0 or 1 — this
        implementation sends one SMTP transaction addressed to all recipients).
        """
        timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
        subject = f"DriveChill Alert: {sensor_name}"
        body = (
            "DriveChill has detected a temperature alert:\n"
            "\n"
            f"Sensor:    {sensor_name}\n"
            f"Value:     {value:.1f}{unit}\n"
            f"Threshold: {threshold:.1f}{unit}\n"
            f"Time:      {timestamp}\n"
            "\n"
            "This alert was sent by DriveChill."
        )
        return await self._send(subject, body)

    async def send_test(self) -> bool:
        """Send a test email to all configured recipients.

        Returns True when at least one message was sent successfully.
        """
        subject = "DriveChill Test Notification"
        body = (
            "This is a test notification from DriveChill. "
            "Your email settings are configured correctly."
        )
        sent = await self._send(subject, body)
        return sent > 0

    async def send_html_report(self, subject: str, html_body: str) -> int:
        """Public API for sending HTML reports (e.g. scheduled analytics)."""
        return await self._send_html(subject, html_body)

    async def _send_html(self, subject: str, html_body: str) -> int:
        """Send an HTML email to all configured recipients.

        Returns 1 on success, 0 on failure or when sending is skipped.
        """
        if aiosmtplib is None:
            logger.warning(
                "aiosmtplib is not installed; email notifications are disabled."
            )
            return 0

        cfg = await self._repo.get()

        if not cfg["enabled"]:
            logger.debug("Email notifications are disabled; skipping send.")
            return 0

        if not cfg["smtp_host"]:
            logger.warning("Email send skipped: smtp_host is not configured.")
            return 0

        recipients: list[str] = cfg["recipient_list"]
        if not recipients:
            logger.warning("Email send skipped: recipient_list is empty.")
            return 0

        password = await self._repo.get_password()

        try:
            async with aiosmtplib.SMTP(
                hostname=cfg["smtp_host"],
                port=cfg["smtp_port"],
                use_tls=cfg["use_ssl"],
                start_tls=cfg["use_tls"] and not cfg["use_ssl"],
            ) as smtp:
                if cfg["smtp_username"] and password:
                    await smtp.login(cfg["smtp_username"], password)

                msg = EmailMessage()
                msg["From"] = cfg["sender_address"]
                msg["To"] = ", ".join(recipients)
                msg["Subject"] = subject
                msg.set_content("This report requires an HTML-capable email client.")
                msg.add_alternative(html_body, subtype="html")
                await smtp.send_message(msg)

            logger.info(
                "HTML email sent: subject=%r recipients=%d", subject, len(recipients)
            )
            return 1

        except Exception:
            logger.exception(
                "Failed to send HTML email: subject=%r smtp_host=%r",
                subject,
                cfg["smtp_host"],
            )
            return 0

    async def _send(self, subject: str, body: str) -> int:
        """Internal: build and send a single SMTP message to all recipients.

        Returns 1 on success, 0 on failure or when sending is skipped.
        """
        if aiosmtplib is None:
            logger.warning(
                "aiosmtplib is not installed; email notifications are disabled. "
                "Install it with: pip install aiosmtplib"
            )
            return 0

        cfg = await self._repo.get()

        if not cfg["enabled"]:
            logger.debug("Email notifications are disabled; skipping send.")
            return 0

        if not cfg["smtp_host"]:
            logger.warning("Email send skipped: smtp_host is not configured.")
            return 0

        recipients: list[str] = cfg["recipient_list"]
        if not recipients:
            logger.warning("Email send skipped: recipient_list is empty.")
            return 0

        password = await self._repo.get_password()

        try:
            async with aiosmtplib.SMTP(
                hostname=cfg["smtp_host"],
                port=cfg["smtp_port"],
                use_tls=cfg["use_ssl"],
                start_tls=cfg["use_tls"] and not cfg["use_ssl"],
            ) as smtp:
                if cfg["smtp_username"] and password:
                    await smtp.login(cfg["smtp_username"], password)

                msg = EmailMessage()
                msg["From"] = cfg["sender_address"]
                msg["To"] = ", ".join(recipients)
                msg["Subject"] = subject
                msg.set_content(body)
                await smtp.send_message(msg)

            logger.info(
                "Email sent: subject=%r recipients=%d", subject, len(recipients)
            )
            return 1

        except Exception:
            logger.exception(
                "Failed to send email: subject=%r smtp_host=%r",
                subject,
                cfg["smtp_host"],
            )
            return 0
